transport/ProbeServer.cpp

```diff
@@ -0,0 +1,224 @@
+    {
+        auto address = _iceConfig.publicIpv4;
+        address.setPort(port);
+        _probeCandidates.emplace_back(ProbeCandidate{ice::TransportType::TCP, address});
+
+        _tcpPorts.insert(port);
+
+        logger::debug("Registered TCP server endpoint %s:%u", _name, address.ipToString().c_str(), port);
+    }
+
+    if (!_iceConfig.publicIpv6.empty() && endpoint.getLocalPort().getFamily() == AF_INET6)
+    {
+        auto address = _iceConfig.publicIpv6;
+        address.setPort(port);
+        _probeCandidates.emplace_back(ProbeCandidate{ice::TransportType::TCP, address});
+
+        _tcpPorts.insert(port);
+
+        logger::debug("Registered TCP server endpoint %s:%u", _name, address.ipToString().c_str(), port);
+    }
+}
+
+void ProbeServer::onServerPortUnregistered(ServerEndpoint& endpoint)
+{
+    // Re-register to keep listening on probes;
+    // TcpServerEndpoint unregisters the listener after dispatching onIceTcpConnect
+    endpoint.registerListener(_probeCredentials.first, this);
+}
+
+void ProbeServer::onIceTcpConnect(std::shared_ptr<Endpoint> endpoint,
+    const SocketAddress& source,
+    const SocketAddress& target,
+    memory::UniquePacket packet)
+{
+    if (endpoint->getTransportType() == ice::TransportType::TCP)
+    {
+        replyStunOk(*endpoint, source, std::move(packet));
+
+        // wait for while to ensure the reply is sent,
+        // before shutting down endpoint
+        u_int32_t gracePeriodMs = 1000;
+        usleep(gracePeriodMs * 1000);
+
+        endpoint->stop(this);
+    }
+}
+
+// Endpoint::IStopEvents
+void ProbeServer::onEndpointStopped(Endpoint* endpoint) {}
+
+void ProbeServer::replyStunOk(Endpoint& endpoint, const SocketAddress& destination, memory::UniquePacket packet)
+{
+    uint64_t timestamp = utils::Time::getAbsoluteTime();
+    const void* data = packet->get();
+
+    auto* stunMessage = ice::StunMessage::fromPtr(data);
+
+    if (stunMessage && stunMessage->isValid() && stunMessage->header.isRequest() &&
+        stunMessage->isAuthentic(_probeCredentials.second))
+    {
+        ice::StunMessage response;
+        response.header.transactionId = stunMessage->header.transactionId;
+        response.header.setMethod(ice::StunHeader::BindingResponse);
+        response.add(ice::StunXorMappedAddress(destination, response.header));
+        response.addMessageIntegrity(_probeCredentials.second);
+        response.addFingerprint();
+
+        endpoint.sendStunTo(destination, response.header.transactionId.get(), &response, response.size(), timestamp);
+    }
+}
+
+// ICE
+const std::pair<std::string, std::string>& ProbeServer::getProbeCredentials() const
+{
+    return _probeCredentials;
+}
+
+const std::vector<ProbeCandidate> ProbeServer::getProbeCandidates()
+{
+    const std::lock_guard<std::mutex> locker(_endpointsLock);
+    return _probeCandidates;
+}
+
+// TODO copied from IceSession.cpp, move to utils/tools and reuse;
```

Comment:
We should make it static in IceSession and let it take the random token as argument.

Reply:
The method generates new tokens for each 10th element (_idGenerator.next()). I guess we should StunTransactionIdGenerator then.